src/jsonpath.js

```diff
@@ -363,6 +363,45 @@ JSONPath.prototype._getPreferredOutput = function (ea) {
         return JSONPath.toPointer(ea.path);
     }
 };
+/**
+ * Detect filter expressions.
+ * @param {string}loc
+ * @returns {boolean}
+ */
+JSONPath.prototype.isFilterExpr = function (loc) {
+    return loc.indexOf('?(') === 0;
+};
+/**
+ * Detects operators in the expression list that require an array result.
+ * an array of results. If no such operator exists, the result
+ * will be treated as a singular value.
+ *
+ * For example, the following paths return whatever is found at the specified
+ * location, whether that is a scalar, object, or array:
+ *   "store.book[0]" - specific book
+ *   "store.bicycle.red" - single property of a single object
+ *
+ * Conversely, the following paths will always result in an array,
+ * because they can generate multiple results depending on the dataset:
+ *   $.store.book[0][category,author] - category,author will return 2 values
+ *   $..book - ".." will recurse through the store object
+ *   $.store.book[1:2] - indicates a range within the array
+ *   $.store.book[*] - wild card indicates multiple results
+ *   $.store.book[?(@.isbn)] - filtering
+ */
+/**
+ * @param {PlainObject} result - json path result
+ * @param {array} exprList - array of json path expressions
+ * @returns {boolean}
+ */
+JSONPath.prototype.isSingularResult = function (result, exprList) {
+    return (result.length === 1 &&
+             !exprList.includes('*') &&
+             !exprList.includes('..') &&
+              exprList.every((loc) => !this.isFilterExpr(loc)) &&
+              exprList.every((loc) => !loc.includes(',')) &&
+              exprList.every((loc) => !loc.includes(':')));
+};
 
```

Comment:
Unless a method is expected to need to reference instance-specific properties, I'd like for us to implement such as these as class methods (e.g., `JSONPath.isFilterExpr` instead of `JSONPath.prototype.isFilterExpr`). This practice allows for using such methods without needing to set up an instance, and also quickly conveys a distinction of purpose.

If you run the `eslint` script, this is one of the issues you'll see reported--as built-in eslint rule [class-methods-use-this](https://eslint.org/docs/rules/class-methods-use-this) (and any others ought to be fixed as well).

However, before making such a change, I am wondering whether it would be more efficient to pass on information about the filter type during iterating (e.g., passing on an `isFilter` boolean similar to what we do with `parent` and `parentPropName` in calls to `_walk`) rather than having to cycle through each expression again?

If possible, I'd also recommend trying with https://github.com/andykais/json-querying-performance-testing to see the impact before and after.

Reply:
@brettz9 good feedback! yes passing a property along while doing the recursion feels much cleaner. I've pushed a commit that takes this approach. Let me know if you'd like any further adjustments.

Next I'm going to do a deep dive on the updated regression tests that @cburgmer just pushed (thanks!), and vet each use case. I'll post my findings.

Reply:
Great, looks very good from my eyeballing... As far as your deep dive on the regression tests, let me know when those tests may only be indicating separate issues, and I can try out and take a final look at this PR. (Though if you find you may be closing to fixing any other issues, we can see about putting them into the same major release, so any other breakage can be handled in one version bump.)

Reply:
@brettz9 I (finally) carved out some time to perform the qa. I went through all the `scalar-consensus` scenarios in the json-path-comparison [regression suite](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml), and validated that (excepting current bugs) `wrap:false` aligns with other libraries. Below are my notes:

| Scalar Rule  | Passes? | Notes  |
| ------------- | ------------- | ------------- |
| [$[2]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L13)  | Yes  |   |
| [$.2](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L21)  | Yes  |   |
| [$[0]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L26)  | Yes  |   |
| [$[-1]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L31)  | No  | Existing issue; see [bug report](https://github.com/cburgmer/json-path-comparison/blob/master/bug_reports/JavaScript_jsonpath-plus.md)   |
| [$['key']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L137)  | Yes  |   |
| [$['@']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L149)  | No  |   Existing issue; see [bug report](https://github.com/cburgmer/json-path-comparison/blob/master/bug_reports/JavaScript_jsonpath-plus.md)|
| [$['$']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L154)  | Yes  |   Existing issue; see [bug report](https://github.com/cburgmer/json-path-comparison/blob/master/bug_reports/JavaScript_jsonpath-plus.md)|
| [$['two.some']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L159)  | Yes  |   
| [$['special:"chars']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L187)  | Yes  |   |
| [$['*']](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L192)  | No  | Existing issue; see [bug report](https://github.com/cburgmer/json-path-comparison/blob/master/bug_reports/JavaScript_jsonpath-plus.md)  |
| [$.key](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L209)  | Yes  |   |
| [$.key](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L214) (on array)  | Yes  |   |
| [$.key](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L219) (on null)  | Yes  |   |
| [$.key](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L219) (on null)  | Yes  |   |
| [$](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L263)  | Yes  |   |

I also reviewed the [test cases](https://github.com/cburgmer/json-path-comparison/blob/JavaScript_jsonpath-plus_nowrap/bug_reports/JavaScript_jsonpath-plus_nowrap.md) generated against the current wrap:false implementation. This confirmed the fixes introduced by this pr, where we're returning a scalar when an array is, in fact, appropriate:

| Scalar Rule  | Passes? | Notes  |
| ------------- | ------------- | ------------- |
| [$[*].bar[*].baz](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L284)  | Yes  |   |
| [$["key"]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L164)  | no consensus  | -  |
| [$[?(@.key)]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L127)  | Yes  |   |
| [$[?(@.key==42)]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L100)  | Yes  |   |
| [$[0:1]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L85)  | Yes  |   |
| [$[0:0]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L81)  | No  |   Existing issue; see [bug report](https://github.com/cburgmer/json-path-comparison/blob/master/bug_reports/JavaScript_jsonpath-plus.md)|
| [$[-1:]](https://github.com/cburgmer/json-path-comparison/blob/master/regression_suite/regression_suite.yaml#L40)  | Yes  |  |

phew! 😅 I'm feeling confident at this point that these changes aligns us with the `scalar consensus` for all other libraries, and also fixes a lot of cases where `wrap:false` is returning an array when it should not 🎉